### PR TITLE
Use fluency instead of fluent-logger-java

### PR DESCRIPTION
**Pull request by repeatedly**
Here are advantages:
- Support multiple instances
- Better network handling
- Flexible buffering

One concern is the max throughput.
fluent-logger-java based:

```
2015-10-06 19:38:21 +0000 [info]: plugin:out_flowcounter_simple count:16304     indicator:num   unit:second
2015-10-06 19:38:22 +0000 [info]: plugin:out_flowcounter_simple count:15975     indicator:num   unit:second
2015-10-06 19:38:23 +0000 [info]: plugin:out_flowcounter_simple count:16540     indicator:num   unit:second
2015-10-06 19:38:24 +0000 [info]: plugin:out_flowcounter_simple count:16681     indicator:num   unit:second
2015-10-06 19:38:25 +0000 [info]: plugin:out_flowcounter_simple count:17097     indicator:num   unit:second
2015-10-06 19:38:26 +0000 [info]: plugin:out_flowcounter_simple count:16330     indicator:num   unit:second
2015-10-06 19:38:27 +0000 [info]: plugin:out_flowcounter_simple count:17377     indicator:num   unit:second
2015-10-06 19:38:28 +0000 [info]: plugin:out_flowcounter_simple count:16304     indicator:num   unit:second
2015-10-06 19:38:29 +0000 [info]: plugin:out_flowcounter_simple count:17213     indicator:num   unit:second
2015-10-06 19:38:30 +0000 [info]: plugin:out_flowcounter_simple count:17265     indicator:num   unit:second
2015-10-06 19:38:31 +0000 [info]: plugin:out_flowcounter_simple count:18483     indicator:num   unit:second
2015-10-06 19:38:32 +0000 [info]: plugin:out_flowcounter_simple count:16497     indicator:num   unit:second
2015-10-06 19:38:33 +0000 [info]: plugin:out_flowcounter_simple count:15986     indicator:num   unit:second
2015-10-06 19:38:34 +0000 [info]: plugin:out_flowcounter_simple count:16361     indicator:num   unit:second
2015-10-06 19:38:35 +0000 [info]: plugin:out_flowcounter_simple count:18290     indicator:num   unit:second
```

fluency based:

```
2015-10-06 19:46:23 +0000 [info]: plugin:out_flowcounter_simple count:6133      indicator:num   unit:second
2015-10-06 19:46:24 +0000 [info]: plugin:out_flowcounter_simple count:6133      indicator:num   unit:second
2015-10-06 19:46:25 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:26 +0000 [info]: plugin:out_flowcounter_simple count:6133      indicator:num   unit:second
2015-10-06 19:46:27 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:28 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:29 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:30 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:31 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:32 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:33 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:34 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:35 +0000 [info]: plugin:out_flowcounter_simple count:12266     indicator:num   unit:second
2015-10-06 19:46:36 +0000 [info]: plugin:out_flowcounter_simple count:18399     indicator:num   unit:second
```

@komamitsu How to improve the performance? stability is good but I want to increase the throughput. Maybe flusher setting?

### CODE REVIEW

**Comment by komamitsu** [comment]
I just took a benchmark roughly on my MBP, but the result was the other way around.
- master branch

```
2015-10-07 13:09:58 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":5925,"kafka.benchtest_bytes":311421,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":5925.0,"kafka.benchtest_bytes_rate":311421.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:09:59 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":31494,"kafka.benchtest_bytes":1724020,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":31494.0,"kafka.benchtest_bytes_rate":1724020.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:00 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":35922,"kafka.benchtest_bytes":1991320,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":35922.0,"kafka.benchtest_bytes_rate":1991320.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:01 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":45803,"kafka.benchtest_bytes":2649059,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":45803.0,"kafka.benchtest_bytes_rate":2649059.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:02 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":55083,"kafka.benchtest_bytes":3249897,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":55083.0,"kafka.benchtest_bytes_rate":3249897.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:03 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":54411,"kafka.benchtest_bytes":3210249,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":54411.0,"kafka.benchtest_bytes_rate":3210249.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:04 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":69778,"kafka.benchtest_bytes":4116902,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":69778.0,"kafka.benchtest_bytes_rate":4116902.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:05 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":71134,"kafka.benchtest_bytes":4196906,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":71134.0,"kafka.benchtest_bytes_rate":4196906.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:06 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":73083,"kafka.benchtest_bytes":4311897,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":73083.0,"kafka.benchtest_bytes_rate":4311897.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:07 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":72581,"kafka.benchtest_bytes":4282279,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":72581.0,"kafka.benchtest_bytes_rate":4282279.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:08 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":73471,"kafka.benchtest_bytes":4334789,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":73471.0,"kafka.benchtest_bytes_rate":4334789.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:09 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":73457,"kafka.benchtest_bytes":4333963,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":73457.0,"kafka.benchtest_bytes_rate":4333963.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:10 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":71606,"kafka.benchtest_bytes":4224754,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":71606.0,"kafka.benchtest_bytes_rate":4224754.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:11 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":72571,"kafka.benchtest_bytes":4281689,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":72571.0,"kafka.benchtest_bytes_rate":4281689.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:12 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":71883,"kafka.benchtest_bytes":4241097,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":71883.0,"kafka.benchtest_bytes_rate":4241097.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:13 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":72947,"kafka.benchtest_bytes":4303873,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":72947.0,"kafka.benchtest_bytes_rate":4303873.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:10:14 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":48851,"kafka.benchtest_bytes":2882209,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":48851.0,"kafka.benchtest_bytes_rate":2882209.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
```
- use-fluency branch

```
2015-10-07 13:12:30 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":134498,"kafka.benchtest_bytes":7581706,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":134498.0,"kafka.benchtest_bytes_rate":7581706.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:12:31 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":129056,"kafka.benchtest_bytes":7614304,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":129056.0,"kafka.benchtest_bytes_rate":7614304.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:12:32 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":64528,"kafka.benchtest_bytes":3807152,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":64528.0,"kafka.benchtest_bytes_rate":3807152.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:12:33 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":193584,"kafka.benchtest_bytes":11421456,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":193584.0,"kafka.benchtest_bytes_rate":11421456.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:12:34 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":64528,"kafka.benchtest_bytes":3807152,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":64528.0,"kafka.benchtest_bytes_rate":3807152.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:12:35 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":129056,"kafka.benchtest_bytes":7614304,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":129056.0,"kafka.benchtest_bytes_rate":7614304.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:12:36 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":129056,"kafka.benchtest_bytes":7614304,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":129056.0,"kafka.benchtest_bytes_rate":7614304.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:12:37 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":129056,"kafka.benchtest_bytes":7614304,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":129056.0,"kafka.benchtest_bytes_rate":7614304.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
2015-10-07 13:12:38 +0900 fluentd.traffic: {"foodb.bartbl_count":0,"foodb.bartbl_bytes":0,"kafka.benchtest_count":26638,"kafka.benchtest_bytes":1571642,"failed_count":0,"failed_bytes":0,"foodb.bartbl_count_rate":0.0,"foodb.bartbl_bytes_rate":0.0,"kafka.benchtest_count_rate":26638.0,"kafka.benchtest_bytes_rate":1571642.0,"failed_count_rate":0.0,"failed_bytes_rate":0.0}
```

I used the following simple script.

```
require 'open3'

Open3.popen3("bin/kafka-console-producer.sh --broker-list localhost:9092 --topic test") do |i, o, e, w|
  1000000.times do |idx|
    i.write %Q${"name":"komamitsu#{idx}","num":#{idx},"email":"komamitsu#{idx}@gmail"}\n$
  end
  i.close
  o.each do |line| p line end
  e.each do |line| p line end
  p w.value
end
```

@repeatedly Can you tell me how you benchmarked?

**Comment by repeatedly** [comment]
@komamitsu Ah, okay. The problem is poseidon based producer's limitation. Using your script, I have similar result.

BTW, could you review my code? Should we provide fluency configuration for users?

File: src/main/java/org/fluentd/kafka/FluentdHandler.java [hunk 1]
```diff
@@ -44,25 +45,36 @@ public void run() {
...
+                    Map<String, Object> data = new HashMap<String, Object>();
+                    data.put("message", new String(entry.message()));
+                    logger.emit("failed", data); // should be configurable
+                }
+            } catch (IOException e) {
+                LOG.error("can't send a log to fluentd. Wait 1 second", e);
+                try {
+                    TimeUnit.SECONDS.sleep(1);
+                } catch (InterruptedException ie) {
+                    LOG.warn("Interrupted during sleep");
```
Comment (komamitsu) [inline src/main/java/org/fluentd/kafka/FluentdHandler.java:76]: call `Thread.currentThread.interrupt()`

File: src/main/java/org/fluentd/kafka/GroupConsumer.java [hunk 2]
```diff
@@ -41,9 +41,8 @@ public GroupConsumer(PropertyConfig config) throws IOException {
...
     }
 
-    public FluentLogger setupFluentdLogger() {
-        final URI uri = config.getFluentdConnect();
-        return FluentLogger.getLogger("", uri.getHost(), uri.getPort());
+    public Fluency setupFluentdLogger() throws IOException {
+        return Fluency.defaultFluency(config.getFluentdConnect());
     }
  
     public void shutdown() {
```
Comment (komamitsu) [inline src/main/java/org/fluentd/kafka/GroupConsumer.java:48]: In the following lines, after the timeout of `executor.awaitTermination()`, call `executor.shutdownNow()`

File: src/main/java/org/fluentd/kafka/PropertyConfig.java [hunk 3]
```diff
@@ -115,16 +116,18 @@ private Properties loadProperties(String propFilePath) throws IOException {
...
+        List<InetSocketAddress> addresses = new ArrayList<InetSocketAddress>();
+
+        for (String address : get(Constants.FLUENTD_CONNECT.key, "localhost:24224").split(",")) {
             try {
-                return new URI("http://localhost:24224/");
-            } catch (URISyntaxException e2) { }
+                URI uri = new URI("http://" + address + "/");
+                addresses.add(new InetSocketAddress(uri.getHost(), uri.getPort()));
+            } catch (Exception e) {
+                throw new RuntimeException("failed to parse '" + address + "' address : " + e.getMessage());
```
Comment (komamitsu) [inline src/main/java/org/fluentd/kafka/PropertyConfig.java]: Maybe you'd better pass the exception as a second parameter like this:

```
 throw new RuntimeException("failed to parse '" + address + "' address ", e);
```

**Comment by komamitsu** [comment]
I left some comments on the changes. Totally looks good.

**Comment by repeatedly** [comment]
@komamitsu Applied! Please re-check!

**Comment by komamitsu** [comment]
LGTM

**Pull request merged**